worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/ContractAdapterImpl.java

```diff
@@ -0,0 +1,186 @@
+                        patientDTO.getDatesUnderContract().add(activeDateRange);
+                    }
+
+                } else {
+                    // new patient id.
+                    // Create a new PatientDTO for this patient
+                    // And then add this month to the patient's datesUnderContract
+
+                    var patientDTO = PatientDTO.builder()
+                            .patientId(bfdPatientId)
+                            .build();
+
+                    if (activeDateRange != null) {
+                        patientDTO.getDatesUnderContract().add(activeDateRange);
+                    }
+
+                    patientDTOs.add(patientDTO);
+                }
+            }
+        }
+
+        return GetPatientsByContractResponse.builder()
+                .contractNumber(contractNumber)
+                .patients(patientDTOs)
+                .build();
+    }
+
+
+    /**
+     * Given a contractNumber and month,
+     * calls BFD api to get a bundle of resources from which the patientIDs are extracted.
+     * If there are multiple pages available, fetches all pages.
+     *
+     * @param contractNumber
+     * @param month
+     * @return a list of PatientIds
+     */
+    private List<String> getPatientIdsForMonth(String contractNumber, Integer month) {
+        Bundle bundle = getBundle(contractNumber, month);
+        final List<String> patientIDs = extractPatientIDs(bundle);
+
+        while (bundle.getLink(Bundle.LINK_NEXT) != null) {
+            bundle = bfdClient.requestNextBundleFromServer(bundle);
+            patientIDs.addAll(extractPatientIDs(bundle));
+        }
+
+        return patientIDs;
+    }
+
+
+    /**
+     * given a contractNumber & a month, calls BFD API to find all patients active in the contract during the month
+     * @param contractNumber
+     * @param month
+     * @return a FHIR bundle of resources containing active patients
+     */
+    private Bundle getBundle(String contractNumber, int month) {
+        try {
+            return bfdClient.requestPartDEnrolleesFromServer(contractNumber, month);
+        } catch (Exception e) {
+            final Throwable rootCause = ExceptionUtils.getRootCause(e);
+            log.error("Error while calling for Contract-2-Bene API : {}", e.getMessage(), rootCause);
+            throw new RuntimeException(rootCause);
+        }
+    }
+
+    /**
+     * Given a Bundle, filters resources of type Patient and returns a list of patientIds
+     * @param bundle
+     * @return a list of patientIds
+     */
+    private List<String> extractPatientIDs(Bundle bundle) {
+        return bundle.getEntry().stream()
+                .map(Bundle.BundleEntryComponent::getResource)
+                .filter(resource -> resource.getResourceType() == ResourceType.Patient)
+                .map(resource -> (Patient) resource)
+                .map(patient -> extractPatientId(patient))
+                .filter(Objects::nonNull)
```

Comment:
We also want to filter our empty strings.

Reply:
done